File: openwebui-core/frontend/openrouter_integration.js
// OpenWebUI Frontend Integration for OpenRouter Gateway
// Add this to your OpenWebUI frontend configuration

// Environment configuration
const OPENROUTER_GATEWAY_URL = process.env.OPENROUTER_GATEWAY_URL || 
    'http://localhost:8000';

// OpenRouter model presets
export const OPENROUTER_PRESETS = [
    {
        id: 'tool-deepseekv3',
        name: 'Tool-DeepSeekV3',
        model: 'deepseek/deepseek-chat',
        description: 'DeepSeek V3 for tool calling and function execution',
        temperature: 0.2,
        capabilities: ['tools', 'function_calling'],
        provider: 'openrouter'
    },
    {
        id: 'vision-glm4v',
        name: 'Vision-GLM4V', 
        model: 'zhipuai/glm-4v-9b',
        description: 'GLM-4V for image analysis and vision tasks',
        temperature: 0.5,
        capabilities: ['vision', 'image_analysis'],
        provider: 'openrouter'
    },
    {
        id: 'venice-uncensored',
        name: 'Venice-Uncensored',
        model: 'venice/uncensored:free',
        description: 'Venice uncensored model for open conversations',
        temperature: 0.8,
        capabilities: ['uncensored', 'creative'],
        provider: 'openrouter'
    },
    {
        id: 'qwen3-coder',
        name: 'Qwen3-Coder',
        model: 'qwen/qwen-2.5-coder-32b-instruct',
        description: 'Qwen 2.5 Coder for programming and development',
        temperature: 0.1,
        capabilities: ['coding', 'programming'],
        provider: 'openrouter'
    }
];

// API client for OpenRouter Gateway
export class OpenRouterGatewayClient {
    constructor(baseUrl = OPENROUTER_GATEWAY_URL) {
        this.baseUrl = baseUrl;
    }

    async chatCompletion(request) {
        const response = await fetch(`${this.baseUrl}/api/v1/chat/completions`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(request)
        });

        if (!response.ok) {
            throw new Error(`Gateway request failed: ${response.status}`);
        }

        return response.json();
    }

    async getModels() {
        const response = await fetch(`${this.baseUrl}/api/v1/models`);
        
        if (!response.ok) {
            throw new Error(`Failed to get models: ${response.status}`);
        }

        return response.json();
    }

    async getHealth() {
        const response = await fetch(`${this.baseUrl}/api/v1/health`);
        
        if (!response.ok) {
            throw new Error(`Health check failed: ${response.status}`);
        }

        return response.json();
    }

    async getRoutingStatus() {
        const response = await fetch(`${this.baseUrl}/api/v1/routing/status`);
        
        if (!response.ok) {
            throw new Error(`Routing status failed: ${response.status}`);
        }

        return response.json();
    }

    async refreshRouting() {
        const response = await fetch(`${this.baseUrl}/api/v1/routing/refresh`, {
            method: 'POST'
        });
        
        if (!response.ok) {
            throw new Error(`Routing refresh failed: ${response.status}`);
        }

        return response.json();
    }
}

// Global client instance
export const gatewayClient = new OpenRouterGatewayClient();

// Model selector integration
export const getOpenRouterModels = async () => {
    try {
        const models = await gatewayClient.getModels();
        
        // Combine with presets for UI display
        const allModels = [
            ...OPENROUTER_PRESETS,
            ...models.openrouter.map(model => ({
                id: model,
                name: model,
                model: model,
                provider: 'openrouter',
                capabilities: []
            })),
            ...models.local_fallback.map(model => ({
                id: `local/${model}`,
                name: `Local: ${model}`,
                model: `local/${model}`,
                provider: 'local_fallback',
                capabilities: ['local', 'offline']
            }))
        ];
        
        return allModels;
    } catch (error) {
        console.error('Failed to load OpenRouter models:', error);
        return OPENROUTER_PRESETS; // Fallback to presets
    }
};

// Chat completion wrapper for OpenWebUI integration
export const sendOpenRouterMessage = async (messages, options = {}) => {
    const request = {
        messages: messages,
        model: options.model,
        temperature: options.temperature || 0.7,
        max_tokens: options.max_tokens || 1200,
        tools: options.tools,
        conversation_id: options.conversation_id,
        stream: options.stream || false
    };

    return gatewayClient.chatCompletion(request);
};

// Health monitoring for UI status indicators
export const checkGatewayHealth = async () => {
    try {
        const health = await gatewayClient.getHealth();
        return {
            status: health.status,
            providers: health.providers,
            routing: health.routing,
            tools: health.tools
        };
    } catch (error) {
        console.error('Gateway health check failed:', error);
        return {
            status: 'unhealthy',
            error: error.message
        };
    }
};

// Routing status for admin interface
export const getRoutingStatus = async () => {
    try {
        return await gatewayClient.getRoutingStatus();
    } catch (error) {
        console.error('Failed to get routing status:', error);
        return { routing_enabled: false, error: error.message };
    }
};

// Utility to get model by preset ID
export const getModelByPreset = (presetId) => {
    return OPENROUTER_PRESETS.find(preset => preset.id === presetId);
};

// Utility to determine if model supports specific capability
export const modelSupportsCapability = (model, capability) => {
    const preset = getModelByPreset(model) || 
        OPENROUTER_PRESETS.find(p => p.model === model);
    
    return preset?.capabilities?.includes(capability) || false;
};

File: openwebui-core/backend/openrouter_integration.py
# OpenWebUI Core Integration for OpenRouter Gateway
# Add this to your OpenWebUI backend configuration

import os
import httpx
from typing import Dict, Any, List, Optional
import logging

logger = logging.getLogger(__name__)

# OpenRouter Gateway Configuration
OPENROUTER_GATEWAY_URL = os.getenv(
    "OPENROUTER_GATEWAY_URL", 
    "http://00-pipelines-gateway:8000"
)
OPENROUTER_GATEWAY_TIMEOUT = int(os.getenv("OPENROUTER_GATEWAY_TIMEOUT", "60"))


class OpenRouterGatewayClient:
    """Client for OpenRouter Gateway integration"""
    
    def __init__(self):
        self.base_url = OPENROUTER_GATEWAY_URL
        self.timeout = OPENROUTER_GATEWAY_TIMEOUT
    
    async def chat_completion(self, 
                             messages: List[Dict[str, str]],
                             model: Optional[str] = None,
                             temperature: float = 0.7,
                             max_tokens: int = 1200,
                             tools: Optional[List[Dict[str, Any]]] = None,
                             conversation_id: Optional[str] = None,
                             stream: bool = False) -> Dict[str, Any]:
        """
        Send chat completion request to OpenRouter Gateway
        """
        payload = {
            "messages": messages,
            "temperature": temperature,
            "max_tokens": max_tokens,
            "stream": stream
        }
        
        if model:
            payload["model"] = model
        if tools:
            payload["tools"] = tools
        if conversation_id:
            payload["conversation_id"] = conversation_id
        
        async with httpx.AsyncClient(timeout=self.timeout) as client:
            try:
                response = await client.post(
                    f"{self.base_url}/api/v1/chat/completions",
                    json=payload,
                    headers={"Content-Type": "application/json"}
                )
                response.raise_for_status()
                return response.json()
                
            except Exception as e:
                logger.error(f"OpenRouter Gateway request failed: {e}")
                raise
    
    async def get_available_models(self) -> Dict[str, Any]:
        """Get available models from OpenRouter Gateway"""
        async with httpx.AsyncClient(timeout=10) as client:
            try:
                response = await client.get(f"{self.base_url}/api/v1/models")
                response.raise_for_status()
                return response.json()
                
            except Exception as e:
                logger.error(f"Failed to get models from gateway: {e}")
                return {"openrouter": [], "local_fallback": []}
    
    async def health_check(self) -> Dict[str, Any]:
        """Check gateway health"""
        async with httpx.AsyncClient(timeout=5) as client:
            try:
                response = await client.get(f"{self.base_url}/api/v1/health")
                response.raise_for_status()
                return response.json()
                
            except Exception as e:
                logger.error(f"Gateway health check failed: {e}")
                return {"status": "unhealthy", "error": str(e)}
    
    async def get_routing_status(self) -> Dict[str, Any]:
        """Get current routing status"""
        async with httpx.AsyncClient(timeout=5) as client:
            try:
                response = await client.get(f"{self.base_url}/api/v1/routing/status")
                response.raise_for_status()
                return response.json()
                
            except Exception as e:
                logger.error(f"Failed to get routing status: {e}")
                return {"routing_enabled": False, "error": str(e)}


# Global gateway client instance
_gateway_client = None


def get_gateway_client() -> OpenRouterGatewayClient:
    """Get or create the global gateway client"""
    global _gateway_client
    if _gateway_client is None:
        _gateway_client = OpenRouterGatewayClient()
    return _gateway_client


# Integration helper functions for OpenWebUI
async def openrouter_chat_completion(messages: List[Dict[str, str]], 
                                    **kwargs) -> Dict[str, Any]:
    """OpenWebUI compatible chat completion function"""
    client = get_gateway_client()
    return await client.chat_completion(messages, **kwargs)


async def get_openrouter_models() -> List[Dict[str, str]]:
    """Get OpenRouter models in OpenWebUI format"""
    client = get_gateway_client()
    models_data = await client.get_available_models()
    
    # Convert to OpenWebUI format
    models = []
    
    # Add OpenRouter models
    for model in models_data.get("openrouter", []):
        models.append({
            "id": model,
            "name": f"OpenRouter: {model}",
            "object": "model",
            "provider": "openrouter"
        })
    
    # Add local fallback
    for model in models_data.get("local_fallback", []):
        models.append({
            "id": f"local/{model}",
            "name": f"Local: {model}",
            "object": "model", 
            "provider": "local_fallback"
        })
    
    return models


async def check_openrouter_gateway_health() -> bool:
    """Check if OpenRouter Gateway is healthy"""
    client = get_gateway_client()
    health = await client.health_check()
    return health.get("status") == "healthy"
